### PR TITLE
MSI/MSI-X Support for Generic Host PCIe Controller

### PR DESCRIPTION
**Pull request by superna9999**
This PR is the second step into PCIe Host Controller support on non-x86 platforms (aka Generic or platform-agnostic) on top of #37668

This adds :

- An API for the GIC ITS to retrieve the MSI base address to be configured in the MSI/MSI-X table
- A generic MSI vector entry when non-x86
- generic controller MSI/MSI-X API and generic core functions for  Generic Host PCIe Controllers
- Implements MSI/MSI-X support for generic controllers & ECAM

Issues with implementation:
- Compliance test fails on: `WARNING:NEW_TYPEDEFS: do not add new typedefs` while adding new type for generic `arch_msi_vector_t` what is the best way to solve that ?

Testing has been done with QEMU with ITS support changes merged for next release.
A branch with ITS is available here:
https://github.com/superna9999/qemu/tree/gicv3-its-v8

### CODE REVIEW

File: include/drivers/pcie/msi.h
Comment (tbursztyka) [inline include/drivers/pcie/msi.h:34]: A comment telling why you don't follow DTS representation (because it chose to follow x86's)

File: include/drivers/pcie/controller.h
Comment (tbursztyka) [inline include/drivers/pcie/controller.h:101]: s/xlate/translate
It's more readable

File: include/drivers/pcie/controller.h
Comment (tbursztyka) [inline include/drivers/pcie/controller.h:256]: s/xlate/translate

File: include/drivers/pcie/controller.h
Comment (tbursztyka) [inline include/drivers/pcie/controller.h:123]: It's weird to see generic_ being put before the domain's prefix. Why not pcie_generic_  ?
Unless this has been already applied like that before.

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h:101]: This should be fixed on #37668 instead

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h:256]: This should be fixed on #37668 instead

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h:123]: This should be fixed on #37668 instead

File: include/drivers/pcie/msi.h
Comment (superna9999) [reply include/drivers/pcie/msi.h:34]: Here it's not about the DT BDF, but the RequesterID format I use to pass to the GIC ITS DeviceID, I took the DEVID name from Linux... but maybe REQID should be better ?

**Comment by AshwinKrishn** [comment]
When will this be merged ? Need it badly for arm64 ivshmem. Thanks.

**Comment by superna9999** [comment]
@AshwinKrishn I'll rebase against `main` asap

**Comment by AshwinKrishn** [comment]
@superna9999 Thank you Neil. Is there a possibility to request owners to request for approval ? This patch will help me with going ahead with my work. Thank you.

**Comment by superna9999** [comment]
@superna9999 patience is key here

@dcpleung would it help if I added myself as code owner of `drivers/pcie/host/controller.c` and `drivers/pcie/host/controller.h` ?

**Comment by dcpleung** [comment]
> @dcpleung would it help if I added myself as code owner of `drivers/pcie/host/controller.c` and `drivers/pcie/host/controller.h` ?

Being a code owner (in the CODEOWNERS file) only affects whether you are being automatically added as a reviewer in PRs. It won't affect the review process.

Also, you need rebase as there are conflicts.

**Comment by superna9999** [comment]
@dcpleung ok, will rebase now, was waiting for #40976 to be merged

**Comment by superna9999** [comment]
@dcpleung do you think it's possible to have this merged before end of merge window close for 3.0.0 ?

**Comment by dcpleung** [comment]
I wish to have input from @jhedberg and @tbursztyka since they are probably impacted by this.

**Review by jhedberg — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

File: include/drivers/pcie/controller.h
Comment (tbursztyka) [reply include/drivers/pcie/controller.h]: Too bad my comment was made the same day #37668 was merged. You don't seem to have made that change.

File: include/drivers/pcie/msi.h
Comment (tbursztyka) [reply include/drivers/pcie/msi.h:34]: keep it as it is.

File: include/drivers/pcie/msi.h
Comment (tbursztyka) [inline include/drivers/pcie/msi.h:21]: msi_vector_generic_t ? 

I already noticed that having generic_ as a prefix is not great. Usually we have the domain first (pcie_ etc...).

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h]: Indeed, let me submit a PR with this name changing

File: include/drivers/pcie/msi.h
Comment (superna9999) [reply include/drivers/pcie/msi.h:21]: OK

File: drivers/pcie/host/msi.c
Comment (tbursztyka) [inline drivers/pcie/host/msi.c:177]: You are already exposing a dedicated msi_vector (generic_msi_vector) as arch_msi_vector_t... shouldn't you consider exposing the arch_pcie_msi_vectors_* functions as well then? 
because now, arch that do are exposing a pcie controller, the generic one, are always going to use the above msi_vector structure anyway.

Comment (superna9999) [reply drivers/pcie/host/msi.c:177]: I can't remeber why I did that, let me try again

Comment (superna9999) [reply drivers/pcie/host/msi.c]: OK, well thanks, I removed an useless commit !

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h]: Submitted as #41949

File: include/drivers/pcie/controller.h
Comment (superna9999) [reply include/drivers/pcie/controller.h]: Submitted as #41948

**Comment by superna9999** [comment]
@dcpleung @jhedberg @tbursztyka Hi, do you have any advice on how we can finalize this PR ?

**Review by carlocaione — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

File: drivers/pcie/host/pcie_ecam.c
Comment (jhedberg) [inline drivers/pcie/host/pcie_ecam.c:343]: Here as well

**Review by jhedberg — CHANGES_REQUESTED** [review]
A couple of minor coding style issues, other wise looks fine to me. @tbursztyka could you also check?

**Review by jhedberg — DISMISSED** [review]

**Review by carlocaione — DISMISSED** [review]

**Review by dcpleung — DISMISSED** [review]

**Comment by jhedberg** [comment]
Seems like this should be ready to be merged, but the "new typedef" failure needs to be overridden by someone with sufficient privileges (@nashif , @carlescufi ?)

**Comment by superna9999** [comment]
Ping @nashif @carlescufi could you help merge this PR ?

**Comment by carlescufi** [comment]
> Ping @nashif @carlescufi could you help merge this PR ?

Working on it. The doc build was failing, I just rebased to see if it gets fixed.

**Comment by jhedberg** [comment]
@carlescufi now it's failing because of a new added typedef:
```
-:289: WARNING:NEW_TYPEDEFS: do not add new typedefs
[24](https://github.com/zephyrproject-rtos/zephyr/runs/5677514780?check_suite_focus=true#step:9:24)
#289: FILE: include/drivers/pcie/msi.h:28:
[25](https://github.com/zephyrproject-rtos/zephyr/runs/5677514780?check_suite_focus=true#step:9:25)
+typedef struct msi_vector_generic arch_msi_vector_t;
```

**Comment by carlescufi** [comment]
> @carlescufi now it's failing because of a new added typedef:
> 
> ```
> -:289: WARNING:NEW_TYPEDEFS: do not add new typedefs
> [24](https://github.com/zephyrproject-rtos/zephyr/runs/5677514780?check_suite_focus=true#step:9:24)
> #289: FILE: include/drivers/pcie/msi.h:28:
> [25](https://github.com/zephyrproject-rtos/zephyr/runs/5677514780?check_suite_focus=true#step:9:25)
> +typedef struct msi_vector_generic arch_msi_vector_t;
> ```

Overriding the Compliance check failure because we do allow a new typedef in this case.

**Pull request merged**
